python_ta/typecheck/base.py

```diff
@@ -219,207 +258,203 @@ def fresh_tvar(self, node: Optional[NodeNG] = None) -> TypeVar:
-            return TypeFail(f'Generic types {_gorg(t1)} and {_gorg(t2)} must have the same number of args')
 
-        unify_result = failable_collect([self.unify(a1, a2) for a1, a2 in zip(t1.__args__, t2.__args__)])
+        g1, g2 = _gorg(conc_tnode1.type), _gorg(conc_tnode2.type)
+        if g1 is not g2 or conc_tnode1.type.__args__ is None or conc_tnode2.type.__args__ is None:
+            # TODO: need to store more info here and in the case below
+            return TypeFail(conc_tnode1, conc_tnode2, ast_node)
+        if len(conc_tnode1.type.__args__) != len(conc_tnode2.type.__args__):
+            return TypeFail(conc_tnode1, conc_tnode2, ast_node)
+
+        unify_result = failable_collect([self.unify(a1, a2) for a1, a2 in
+                                         zip(conc_tnode1.type.__args__,
+                                             conc_tnode2.type.__args__)])
         if isinstance(unify_result, TypeFail):
             return unify_result
         unified_args = unify_result.getValue()
 
-        if g1 == Tuple:
-            tuple_args = tuple(unified_args)
-            # Handle the special case when t1 or t2 are empty tuples; TODO: investigate this
-            if tuple_args == ((),):
-                tuple_args = ()
-            return TypeInfo(Tuple[tuple_args])
-        elif g1 == Callable:
-            return TypeInfo(Callable[unified_args[:-1], unified_args[-1]])
-        else:
-            return TypeInfo(g1[tuple(unified_args)])
-
-    def _merge_sets(self, t1: TypeVar, t2: TypeVar) -> TypeResult:
-        """Merge the two sets that t1 and t2 belong to.
-
-        TODO: make sure it's OK to return a TypeResult and not an exception
-        """
-        # TODO: look into implementation of  __eq__ for TypeVar to make sure we can use == here.
-        if t1 == t2:
-            return TypeInfo(t1)
-
-        rep1 = self._find(t1)
-        rep2 = self._find(t2)
-        if isinstance(rep1.type, TypeVar) and isinstance(rep2.type, TypeVar):
-            if rep1.type.__name__ < rep2.type.__name__:
-                rep2.parent = rep1
-                return TypeInfo(rep1.type)
-            else:
-                rep1.parent = rep2
-                return TypeInfo(rep2.type)
-        elif isinstance(rep2.type, TypeVar):
-            rep2.parent = rep1
-            return TypeInfo(rep2.type)
-        elif isinstance(rep1.type, TypeVar):
-            rep1.parent = rep2
-            return TypeInfo(rep2.type)
-        else:
-            # In this case both set representatives are concrete types.
-            # If they're compatible, we can still unify the sets. Otherwise, an error
-            # is raised here.
-            if rep1.type == rep2.type:
-                rep2.parent = rep1
-                return TypeInfo(rep1.type)
-            else:
-                return TypeFail(f'Incompatible types {rep1.type} and {rep2.type}')
+        # TODO: should the edges instead be between the original _TNodes?
```

Comment:
Yes, this strikes me as odd. I'm not sure why this is better than connecting the original nodes.